### PR TITLE
[DX][Security] Added TokenStorage::getUser($userClass)

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

---

This is a PR to get **feedback on the feature not the code**.
There a more thing to do so please don't comment on the code.

---

So why this feature?

Actually I'm facing an issue right now and I don't really know what is the best
solution.

Technically:

* `$token = $this->tokenStorage->getToken()` could return `null` or a `TokenInterface`;
* `$user = $token->getUser()` could return a `UserInterface` or a `string` or an object with `__toString()`;
* even if `$user instanceOf UserInterface` it can be so any implementation of `UserInterface`. For example `App\Entity\User` or `App\Entity\Agent`.

So calling somewhere `$this->tokenStorage->getToken()->getUser()->getFoobar()`
could lead to many fatal errors.

A first solution could be to over-protect the code, like it's done here: https://github.com/symfony/symfony/blob/07766b39052b6ba294dc6ddf85e38ee61a8e0bcb/src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php#L352-L361
And an extra if should be added to ensure the class of `UserInterface`.
So we should write:

```php
public function doSomething()
{
    $user = $this->getUser();
    if (!$user) {
        return; // or exception...
    }

    dump($user->getFoobar());
}

private getUser()
{
    if (!$this->tokenStorage->getToken()) {
        return null;
    }

    $user = $token->getUser();

    if (!$user) {
        return null;
    }

    if (!$user instanceof User) {
        return null;
    }

    return $user;
}
```

But it's a bit boring... And it's even more boring if we really know that this
call can be called if and only if the `user` firewall has been used.  So we are
100% sure no fatal error could be thrown... But only for now. May be in a soon
future, someone will change the firewall. And BOOM.

So Basically, I don't know what is the best thing to do:

1/ Be really robust and write lot of code
2/ Be pragmatic

But, I would like to propose another solution... This PR ;)

WDYT?

### CODE REVIEW

**Comment by lyrixx** [comment]
> This looks unintuitive -> TokenStorage::getUser() you are getting the user from something that stores tokens, not users.

You are right about this. More over Adding such method will be complicated because the method will not be in the Interface... 

May be we can introduce a `UserStorageInterface` instead. I will increase the DX by hiding the "Token stuff".

**Comment by mvrhov** [comment]
I'm against this. We should decouple the Security user from the real user like described by @iltar in [his article](https://stovepipe.systems/post/decoupling-your-security-user)

**Comment by lyrixx** [comment]
@mvrhov Thanks for the Article. But actually, both approaches are fully compatible and not related.

**Comment by ro0NL** [comment]
> But it's a bit boring... 

Do it once :) IMHO the trick is to create service for your approach (which might fit core) to utilize the TokenStorage usage, i.e. dont rely on such a low level API directly.

Here's [my implementation](https://github.com/msgphp/msgphp/blob/master/src/User/Infra/Security/SecurityUserFactory.php) of such a service today. Yes it's tied to my UserInterface implementation and does some conversion to another entity object (as mentioned in above article), but that's not really related indeed.

So i inject `SecurityUserFactory $factory` call `$factory->getUser()` and im done. (in my case i have the entity for the current user not, not the UserInterface).

Another benefit is i throw the corresponding security exceptions, meaning those are picked up by the listener stuff (e.g. redirect to login). Besides it saves a lot of boring null checks :-)

**Comment by chalasr** [comment]
Same feeling as others, token storage is a low level api which does not know about the token api itself, which is good IMO. We already have a `getUser()` shortcut in [the `security` service](https://github.com/symfony/security/blob/master/Core/Security.php#L38) and in the controller layer.

> both approaches are fully compatible and not related.

IIUC the main goal of this feature is to get an instance of a specific user class. Can you give a use case where you have multiple security users (UserInterface implementations)? 
I don't have one, the difference relates to domain to me, i.e. `$this->tokenStorage->getUser()->getDomainUser() instance of Agent` or `->isAgent()` etc..

**Comment by lyrixx** [comment]
> Same feeling as others, token storage is a low level api which does not know about the token api itself, which is good IMO. We already have a getUser() shortcut in the security service and in the controller layer.

Oh, I forgot the the `security` service ;)



> IIUC the main goal of this feature is to get an instance of a specific user class. Can you give a use case where you have multiple security users (UserInterface implementations)?

Sure. The first one is a regular user. Authenticated via oAuth (for example). Another "user" is an Environment (I replaced Agent by Environment here to make it simple). An Environment is authenticated via http basic for exemple. For example a similar system is used on Blackfire where you can login, and your Environment are logged also. So basically, `$this->getUser` could return a User instance, or and Environment instance.

**Comment by ostrolucky** [comment]
Yeah that Security helper is really nice, didn't know about it. 

Writing service for retrieving current domain user is _really_ common and boring, but that helper reduces such task into this service in user space, which is optimal solution, since it's static analysis friendly and doesn't force you to specify FCQN each time you want to retrieve non-default user object.

```php
class CurrentUserStorage {
  /** @var Security */
  public $security;

  public function getUser(): ?User
  {
     return $user = $this->security->getUser() and $user instanceof User ? $user : null;
  }

  public function getEnvironment(): ?Environment
  {
       return $user = $this->security->getUser() and $user instanceof Environment ? $user : null;
  }
}
```

That said, adding optional argument to `Security::getUser` doesn't cost much and would help to reduce above into
```php
class CurrentUserStorage {
  /** @var Security */
  public $security;

  public function getUser(): ?User
  {
     return $this->security->getUser(User::class);
  }

  public function getEnvironment(): ?Environment
  {
       return $this->security->getUser(Environment::class);
  }
}
```

**Review by Simperfit — APPROVED** [review]

**Comment by lyrixx** [comment]
@javiereguiluz what do you think about merging this feature to the the [`Security`](https://github.com/symfony/security/blob/master/Core/Security.php#L38) class ?

**Comment by iltar** [comment]
> Writing service for retrieving current domain user is really common and boring

And a domain user has no place in a framework, so 👎 from me for this PR. If you need the UserInterface, you'll have to ask for it in your controller via `UserInterface $user` in your method arguments, which does exactly this. If you need it in any other place in the framework, you'll have to ensure the security component is supporting this location (think event listeners, but this is a _bad_ place). If you don't actually need the security user, which you probably don't as you need your domain user, you should probably add an argument value resolver for your domain user.

Perhaps this part could be made a bit easier in Symfony? Symfony can't provide you with domain objects, but it _can_ make it easier to fetch them.

```yaml
# security.yaml
security:
    domain_users:
        App\Security\FooUser: App\Domain\FooUser
        App\Security\BarUser: App\Domain\BarUser
        App\Security\OtherUsersInterface: App\Domain\GenericUser
```

Then the argument value resolver could be resulting in something like:

```php
public function fooAction(FooUser $user);
public function barAction(BarUser$user);
public function genericAction(GenericUser $user);
```

Note that if those users are entities, it _will_ conflict with the doctrine parameter converter (which should probably die in a fire anyway).

The primary advantage here, is that we don't make the developers rely on the Symfony security internals. This means that if (for example) we decide that we want to make the `UserInterface` something internal, like only used during the authentication process but never exposed to the application, it can be done without having to refurbish the entire application with new arguments.

I've been personally using this strategy and works like a charm in my application. I can change whatever I want in my security config, because my actions only want the `Client` object, not a security user. If your Security user would've been the Domain user (not a good idea, but perfectly legit), this would still work!

**Comment by lyrixx** [comment]
> If you need it in any other place in the framework, you'll have to ensure the security component is supporting this location (think event listeners, but this is a bad place)

This is how it works in API Platform. (This is why I opened this PR)

> it will conflict with the doctrine parameter converter (which should probably die in a fire anyway).

Why should it die in fire ?

**Comment by iltar** [comment]
>> If you need it in any other place in the framework, you'll have to ensure the security component is supporting this location (think event listeners, but this is a bad place)
> This is how it works in API Platform. (This is why I opened this PR)

Currently the HTTP supports the UserInterface as value resolver, which is why you don't need it elsewhere, the security is not doing anything else. Also, API platform is HTTP, so it can make use of the resolver. If you _do_ have a listener (updated by for example) and you do not follow the best practices and you _still_ want to rely on this kind of behavior, you can already do `$tokenStorage->getToken()->getUsername();` and check if that's not null. If you don't have a token, this listener shouldn't have been called anyway. If you don't have a compatible user object, you will still have to check this in the listener.

I've had a couple of listeners doing things like that and it _always_ causes trouble. Think of command line scripts that modify this entity that don't have security and thus fail. Think of listeners that for some reason aren't triggered, are triggered too late or triggered wrongly where you didn't expect it.

To me it sounds like this PR is meant to couple the domain and framework, which is something that should not be done imo.

>> it will conflict with the doctrine parameter converter (which should probably die in a fire anyway).
> Why should it die in fire ?

Because it has terrible side effects. If your first argument is an entity, it will also look at the second argument and use it to query, even if this argument is 100% unrelated to your entity (but happens to share the name). I don't want to go off-topic about this, it's just something that I've turned off the second I could.

**Comment by ostrolucky** [comment]
> And a domain user has no place in a framework, so 👎 from me for this PR

This PR has no dependency on domain user

> Perhaps this part could be made a bit easier in Symfony? Symfony can't provide you with domain objects, but it can make it easier to fetch them.

This is what is this PR about

**Comment by iltar** [comment]
>> And a domain user has no place in a framework, so 👎 from me for this PR
>
> This PR has no dependency on domain user

This method creates a hard dependency on the user. You will only need the identifier of this user (username) and you can already obtain this from the token. This gives me the idea that the User object will be used for more than the username, it'll be used like the domain object, making the framework suddenly return a domain object,



> Perhaps this part could be made a bit easier in Symfony? Symfony can't provide you with domain objects, but it can make it easier to fetch them.
>
> This is what is this PR about

No, this PR adds a method that returns a domain object, which btw, can't ever be properly type-hinted.

**Comment by lyrixx** [comment]
> I've had a couple of listeners doing things like that and it always causes trouble. Think of command line scripts that modify this entity that don't have security and thus fail. Think of listeners that for some reason aren't triggered, are triggered too late or triggered wrongly where you didn't expect it.

Actually, API platform listener only listen kernel.* events. So there are no issues with CLI.

> This method creates a hard dependency on the user. You will only need the identifier of this user (username) and you can already obtain this from the token. This gives me the idea that the User object will be used for more than the username, it'll be used like the domain object, making the framework suddenly return a domain object,

Actually, It's not an issue for me. I understand that for **some** application it could be wrong, but for many application, it **is not** an issue. I have built / maintained SF application for year and it has never be an issue **to me**. But I totally understand your POV and respect it.

---

Anyway I think we are going a bit off-topic here. As I [said](https://github.com/symfony/symfony/pull/25617#issuecomment-354294109) I totally forgot the Security Class. So basically, now I just want to add 3 LOC in it now.

I will update this PR according to.

**Review by chalasr — COMMENTED** [review]
This makes `Security::getUser()` returns null if the user is not an instance of the specified class, implicitly promoting to couple domain logic to the framework (different security users). 
👎  for me, better keep solving that in userland via `instanceof` checks.

File: src/Symfony/Component/Security/Core/Security.php [hunk 1]
```diff
@@ -35,7 +35,7 @@ public function __construct(ContainerInterface $container)
     /**
      * @return UserInterface|null
      */
-    public function getUser()
+    public function getUser(string $userClass)
```
[the remark this answers is not in the document]
Comment (lyrixx) [reply src/Symfony/Component/Security/Core/Security.php]: Of course. This is a typo :-P

**Comment by lyrixx** [comment]
Closing as many people are against this PR.
